File: examples/clean-api-python/application/use_cases/create_product.py
"""Create product use case - Application layer"""
from dataclasses import dataclass
from typing import Dict, Any
import sys
import os

# Add backbone to path
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../..')))

from backbone.application.exceptions import ValidationException
from backbone.infrastructure.logging import LoggerFactory, LogLevel
from examples.clean_api_python.domain.entities.product import Product, ValidationError
from examples.clean_api_python.domain.repositories.product_repository import ProductRepository
import time


@dataclass
class CreateProductInput:
    """Input for creating a product"""
    name: str
    description: str
    price: float
    category: str
    stock: int


@dataclass
class CreateProductOutput:
    """Output of product creation"""
    product: Product


class CreateProductUseCase:
    """Handles product creation"""
    
    def __init__(self, repository: ProductRepository):
        self.repository = repository
        self.logger = LoggerFactory.create_layer_logger(
            "product-api",
            "application",
            "CreateProductUseCase"
        )
    
    def execute(self, input: CreateProductInput, context: Dict[str, Any]) -> CreateProductOutput:
        """Executes the use case"""
        self.logger.log(
            LogLevel.INFO,
            "Creating product",
            extra_data={
                "name": input.name,
                "category": input.category,
                "price": input.price
            },
            context=context
        )
        
        # Validar input
        self._validate_input(input)
        
        # Crear entidad de dominio
        try:
            product = Product(
                name=input.name,
                description=input.description,
                category=input.category,
                price=input.price,
                stock=input.stock
            )
        except ValidationError as e:
            self.logger.log(
                LogLevel.ERROR,
                "Domain validation failed",
                extra_data={
                    "error": str(e),
                    "field": e.field,
                    "code": e.code
                },
                context=context
            )
            raise ValidationException(
                "Product validation failed",
                [{"field": e.field, "message": e.message}]
            )
        
        # Guardar en repositorio
        start = time.time()
        try:
            self.repository.create(product)
        except Exception as e:
            duration_ms = int((time.time() - start) * 1000)
            self.logger.log(
                LogLevel.ERROR,
                "Failed to save product",
                extra_data={
                    "error": str(e),
                    "duration_ms": duration_ms,
                    "error_code": 10001003
                },
                context=context
            )
            raise Exception(f"Failed to create product: {str(e)}")
        
        duration_ms = int((time.time() - start) * 1000)
        self.logger.log(
            LogLevel.INFO,
            "Product created successfully",
            extra_data={
                "product_id": product.id,
                "duration_ms": duration_ms
            },
            context=context
        )
        
        return CreateProductOutput(product=product)
    
    def _validate_input(self, input: CreateProductInput) -> None:
        """Validates input"""
        if not input.name:
            raise ValueError("name is required")
        if not input.category:
            raise ValueError("category is required")
        if input.price <= 0:
            raise ValueError("price must be greater than 0")
        if input.stock < 0:
            raise ValueError("stock cannot be negative")

File: backbone-go/examples/clean-api-go/application/usecases/create_product.go
// Package usecases contains application use cases
package usecases

import (
	"context"
	"fmt"
	"time"

	"github.com/freakjazz/backbone-go/application/exceptions"
	"github.com/freakjazz/backbone-go/examples/clean-api-go/domain/entities"
	"github.com/freakjazz/backbone-go/examples/clean-api-go/domain/repositories"
	"github.com/freakjazz/backbone-go/infrastructure/logging"
)

// CreateProductInput represents the input for creating a product
type CreateProductInput struct {
	Name        string  `json:"name"`
	Description string  `json:"description"`
	Price       float64 `json:"price"`
	Category    string  `json:"category"`
	Stock       int     `json:"stock"`
}

// CreateProductOutput represents the output of product creation
type CreateProductOutput struct {
	Product *entities.Product `json:"product"`
}

// CreateProductUseCase handles product creation
type CreateProductUseCase struct {
	repository repositories.ProductRepository
	logger     *logging.EnhancedLogger
}

// NewCreateProductUseCase creates a new use case instance
func NewCreateProductUseCase(repository repositories.ProductRepository, logger *logging.EnhancedLogger) *CreateProductUseCase {
	return &CreateProductUseCase{
		repository: repository,
		logger:     logger,
	}
}

// Execute executes the use case
func (uc *CreateProductUseCase) Execute(ctx context.Context, input CreateProductInput) (*CreateProductOutput, error) {
	// Logger con contexto de aplicación
	useCaseLogger := uc.logger.
		WithLayer("application").
		WithHandler("CreateProductUseCase").
		WithMethod("Execute").
		WithContext(map[string]interface{}{
			"request_id": ctx.Value("request_id"),
			"user_id":    ctx.Value("user_id"),
		})

	useCaseLogger.Info("Creating product", map[string]interface{}{
		"name":     input.Name,
		"category": input.Category,
		"price":    input.Price,
	})

	// Validar input
	if err := uc.validateInput(input); err != nil {
		useCaseLogger.ErrorWithCode("Invalid input", 10001001, map[string]interface{}{
			"error": err.Error(),
		})
		return nil, err
	}

	// Crear entidad de dominio
	product, err := entities.NewProduct(
		input.Name,
		input.Description,
		input.Category,
		input.Price,
		input.Stock,
	)
	if err != nil {
		useCaseLogger.ErrorWithCode("Domain validation failed", 10001002, map[string]interface{}{
			"error": err.Error(),
		})
		return nil, exceptions.NewValidationException(
			"Product validation failed",
			[]exceptions.ValidationError{
				{Field: "product", Message: err.Error()},
			},
		)
	}

	// Guardar en repositorio
	start := time.Now()
	if err := uc.repository.Create(ctx, product); err != nil {
		duration := time.Since(start).Milliseconds()
		useCaseLogger.ErrorWithCode("Failed to save product", 10001003, map[string]interface{}{
			"error":       err.Error(),
			"duration_ms": duration,
		})
		return nil, exceptions.NewUseCaseException(
			"Failed to create product",
			"CreateProductUseCase",
		)
	}

	duration := time.Since(start).Milliseconds()
	useCaseLogger.Info("Product created successfully", map[string]interface{}{
		"product_id":  product.ID,
		"duration_ms": duration,
	})

	return &CreateProductOutput{
		Product: product,
	}, nil
}

func (uc *CreateProductUseCase) validateInput(input CreateProductInput) error {
	if input.Name == "" {
		return fmt.Errorf("name is required")
	}
	if input.Category == "" {
		return fmt.Errorf("category is required")
	}
	if input.Price <= 0 {
		return fmt.Errorf("price must be greater than 0")
	}
	if input.Stock < 0 {
		return fmt.Errorf("stock cannot be negative")
	}
	return nil
}
